VEGABoard BLE SW LL

alexandru-porosanu-nxp opened the pull request:
This PR adds support for BLE on VEGABoard

Currently the following sample applications are supported by the BLE SW LL
and BLE host stack running on RI5CY core of RV32M1 SoC:

- BEACON
- CENTRAL
- CENTRAL_HR
- EDDYSTONE
- HCI_UART
- IBEACON
- PERIPHERAL_CSC (Cycling Speed Cadence)
- PERIPHERAL_DIS (Device Information Service)
- PERIPHERAL_ESP (Environmental Sensing Service)
- PERIPHERAL_HR (Heart Rate)
- PERIPHERAL_HT (Health Thermometer)
- SCAN_ADV

Limitations:
- no 2 Mbps, 512/256 Kbps PHY
- no controller-based cryptographic services
- no mesh support
- no power-save

This PR depends on https://github.com/zephyrproject-rtos/hal_openisa/pull/1 being merged

George-Stefan committed:
drivers: entropy: add openisa/RV32M1 entropy driver

Wrapper for openisa/RV32M1 TRNG driver

Signed-off-by: George Stefan <george.stefan@nxp.com>

alexandru-porosanu-nxp committed:
dts: riscv: add Generic FSK node

The Generic FSK controller enables radio operation
using a custom GFSK/GMSK or MSK modulation format
achieved by programming a set of PHY variables such
as BT product, modulation index and modulation filter
co-efficients (such that max frequency deviation
<= 500kHz). Generic FSK mode also offers a highly
configurable packet structure, variable bit rate
transmission and reception, some limited packet
(header) processing, and interface to a RAM-based
Packet Buffer.

Signed-off-by: Alex Porosanu <alexandru.porosanu@nxp.com>

alexandru-porosanu-nxp committed:
dts: bindings: add modem binding for RV32M1 Generic FSK node

The Generic FSK controller enables radio operation
using a custom GFSK/GMSK or MSK modulation format
achieved by programming a set of PHY variables such
as BT product

This binding describes the devices and its associated properties.

Signed-off-by: Alex Porosanu <alexandru.porosanu@nxp.com>

Commit:
boards: riscv: rv32m1: use xoroshiro as default RNG

Don't use use the RV32M1 TRNG as a random source since it can
be  quite slow. Instead, use the software implemented xoroshiro
RNG.

Signed-off-by: David Leach <david.leach@nxp.com>

Commit:
soc: riscv: rv32m1: kconfig: use RV32M1 TRNG as an entropy generator

If the user requires an entropy generator to be activated, enable
the SoC TRNG as a source for the entropy.

Signed-off-by: David Leach <david.leach@nxp.com>

Commit:
boards: riscv: rv32m1: enable BT related configuration

Specific SW defined BLE LL parameters need to be set
if the user enables it on this platform. As such, conditionally
enable them directly into the defconfig.

INTMUX CH2 and CH3 are not available to be used if BT support
is enabled on Vega, because they are used internally by the
BLE SW LL

Signed-off-by: David Leach <david.leach@nxp.com>

Commit:
boards: riscv: rv32m1: add support for GPIO debugging

Having a pin toggle when the code reaches a certain point
is really useful for debugging; the infrastructure is already
in place for Nordic boards, so just build upon and enable the
mechanism on the Vega board as well.

Signed-off-by: David Leach <david.leach@nxp.com>

alexandru-porosanu-nxp committed:
samples: bluetooth: add overlays for RV32M1 SoC for BLW SW LL

Add the required specific HW configuration for SW defined
BLE LL on RV32M1 SoC, by means of DTS overlays:
- enable INTMUX0 channels 2 & 3
- route Generic FSK RF0 interrupt to INTMUX channel 3
- route LPTMR1 interrupt to INTMUX channel 2

This change is done for all Bluetooth samples that are intended to
run on RV32M1 using BLE SW LL.

Signed-off-by: George Stefan <george.stefan@nxp.com>
Signed-off-by: Radu Alexe <radu.alexe@nxp.com>
Signed-off-by: Ionut Ursescu <ionut.ursescu@nxp.com>
Signed-off-by: Alex Porosanu <alexandru.porosanu@nxp.com>

George-Stefan committed:
bluetooth: controller: openisa/RV32M1: add HAL for LPTMR & GenFSK

This commit adds the HAL layer needed for the BLE controller
LL on RV32M1 SoCs on OpenISA boards. Specifically, the controller
makes use of the the GenFSK and LPTMR IP blocks.

Signed-off-by: George Stefan <george.stefan@nxp.com>
Signed-off-by: Radu Alexe <radu.alexe@nxp.com>
Signed-off-by: Ionut Ursescu <ionut.ursescu@nxp.com>

George-Stefan committed:
bluetooth: controller: openisa/RV32M1: BLE Link Layer ULL/LLL split

This commit takes the Nordic LLL and adapts it for RV32M1 SoCs, using
the blocks that are specific to this SoC: the GenFSK & LPTMR IP
blocks.

Signed-off-by: George Stefan <george.stefan@nxp.com>
Signed-off-by: Radu Alexe <radu.alexe@nxp.com>
Signed-off-by: Ionut Ursescu <ionut.ursescu@nxp.com>

George-Stefan committed:
bluetooth: controller: openisa/RV32M1: enable the BLE SW controller

This patch updates the config files required to enable the
BLE SW defined controller to be built on RV32M1 SoCs. Only the split
version is supported.

Signed-off-by: George Stefan <george.stefan@nxp.com>
Signed-off-by: Radu Alexe <radu.alexe@nxp.com>
Signed-off-by: Ionut Ursescu <ionut.ursescu@nxp.com>

Commit:
bluetooth: controller: openisa/RV32M1: add debug pins support

Add SW defined BLE LL debug support on Vega platform by using
the debug GPIO pins infrastructure.

Signed-off-by: David Leach <david.leach@nxp.com>

alexandru-porosanu-nxp committed:
doc: riscv: rv32m1_vega: add BLE software link layer info

Since the experimental BLE software link layer is enabled on
the VEGABoard, add some information about it, as well as the
limitations.

Signed-off-by: Alex Porosanu <alexandru.porosanu@nxp.com>

cvinayak commented on boards/riscv/rv32m1_vega/Kconfig.defconfig:
```diff
@@ -68,4 +69,80 @@ config I2C_3
+config BT_CTLR_ADV_EXT
+	default n
+# Odd one to turn on?
+config BT_CTLR_ADVANCED_FEATURES
```
nRF specific Kconfig could be under depends. But this can be done as port of new PR

alexandru-porosanu-nxp replied:
OK, we can do it a new PR

cvinayak commented on samples/bluetooth/beacon/rv32m1_vega_ri5cy.overlay:
```diff
@@ -0,0 +1,22 @@
+/*
```
can these overlay be in some common folder for the samples?

alexandru-porosanu-nxp replied:
I think I can do a couple of changes to the build/overlay system so the overlays for a platform are taken from some other directory (rather than the sample directory).
I'm thinking of having a per-platform/per SoC directory structure like for instance overlays/riscv/rv32m1/rv32m1_vega_ri5cy.overlay, which would get overlayed only if BT=y.
This PR is not the best container for such a change, as such I'd create a separate one that allows the overlays to be decoupled from the sample dir. Does this make se?

cvinayak replied:
this may be more a generic topic, I have no strong objection. May be @jhedberg has some comments?

dleach02 replied:
@galak, This was the overlays I was talking to you about re: DTS. Can you provide some guidance here?

cvinayak commented on subsys/bluetooth/controller/hci/openisa/hci_vendor.h:
```diff
@@ -0,0 +1,15 @@
+/*
+ * Copyright (c) 2016 Nordic Semiconductor ASA
```
copyright correction

alexandru-porosanu-nxp replied:
Our internal copyright rules say that (C) should never be removed from a file. If a file is significantly changed, then NXP copyright can be added. Since this file is not really changed, I didn't see the need to update the copyright.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ctrl.c:
```diff
@@ -4895,10 +4915,12 @@ static inline void isr_radio_state_close(void)
 	event_inactive(0, 0, 0, NULL);
 
+#if defined(CONFIG_SOC_COMPATIBLE_NRF)
 	err = clock_control_off(_radio.hf_clock, NULL);
```
have ```int err = ``` and remove the cond, compile above in the beginning of the function. Similarly at all other places.

alexandru-porosanu-nxp replied:
OK, that's doable

cvinayak commented on subsys/bluetooth/controller/ll_sw/ctrl.c:
```diff
@@ -1,6 +1,7 @@
 /*
```
may be create new PR for legacy controller changes?

alexandru-porosanu-nxp replied:
I can take the relevant commit out of this PR and add it to a new one, but I'd rather have everything Vega SW LL related into one PR. There's also the option of dropping the legacy support altogether. What do you think?

cvinayak replied:
If the legacy controller is working then I have no objection to keep it. If its not too much work to keep it in separate commit, please do. It will be easy in future in case doing a bisect or to revert just legacy controller file changes.

aescolar replied:
@alexandru-porosanu-nxp Given that the legacy controller is meant to be deprecated in around 1 release, it may be just simpler to drop its changes, and therefore minimize the amount of extra work.

alexandru-porosanu-nxp replied:
OK then, I will drop the corresponding commit from the PR

dleach02 replied:
My suggestion was to remove legacy controller changes for the exact same reason @aescolar made. To me it simplifies the code to remove the unnecessary conditional compile blocks.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull.c:
why not have this define in vendor hal headers, and why the global variable use?

George-Stefan replied:
It's used in radio.c, where rx buffer size is checked against PDU size

George-Stefan replied:
A function in ULL returning the Rx buffer size is better?

cvinayak replied:
can you use the max_len passed in radio_pkt_configure call? then you dont need this global.

George-Stefan replied:
done

cvinayak commented on subsys/bluetooth/controller/ll_sw/openisa/hal/cpu.h:
```diff
@@ -0,0 +1,13 @@
+static inline void cpu_sleep(void)
+{
+	__WFE();
+	/* __SEV(); */
```
__SEV is used to ensure second __WFE does not wait. Assume the function entered with event flag set, then first __WFE will clear and pass through, without __SEV() the second __WFE will wait.

George-Stefan replied:
ext/hal/openisa/vega_sdk_riscv/RISCV/core_riscv32.h
```
__attribute__((always_inline)) __STATIC_INLINE void __WFE(void)
{
}

```

alexandru-porosanu-nxp replied:
We'll have a look at this, and see how we can get a similar behavior as for ARM cores

cvinayak replied:
sorry... my fault... of course these instructions are not applicable for RISCV

cvinayak reviewed (changes requested):
Did a quick pass, as most of the implementation is in the vendor specific parts I will leave it for others to review.

alexandru-porosanu-nxp commented on boards/riscv/rv32m1_vega/Kconfig.defconfig:
```diff
@@ -68,4 +69,80 @@ config I2C_3
+config BT_CTLR
+	default y
+
+choice BT_LL_CHOICE
```
OK, so we can remove this

aescolar commented on subsys/bluetooth/controller/ll_sw/openisa/README:
```diff
@@ -0,0 +1,23 @@
+
+Limitations:
+- no 2 Mbps, 512/256 Kbps PHY
+- no controller-based cryptographic services
```
@alexandru-porosanu-nxp It seems it is not only private address resolution that is not supported, but also LL packet encryption? (if so maybe it would be worth mentioning it explicitly as the 1st one is a relatively advanced feature while the 2nd one is quite basic)

alexandru-porosanu-nxp replied:
Nope, neither RPA nor LL packet encryption are supported for now. I will explicitly state this.

aescolar commented on subsys/bluetooth/controller/ll_sw/openisa/hal/RV32M1/radio/radio.c:
```diff
@@ -0,0 +1,1009 @@
+	 * undefined behavior.
+	 * Those value have an associated db level that was not found in the
+	 * documentation.
+	 * Because of these inconsistencies for the moment this
```
Seems this would be worth adding to the list of not yet supported features

alexandru-porosanu-nxp replied:
Right, will add this as well to the list of known limitations

aescolar commented on boards/riscv/rv32m1_vega/Kconfig.defconfig:
```diff
@@ -68,4 +69,80 @@ config I2C_3
+
+# Yep, we don't have HW acceleration for TRX switching. But this is very much
+# a NRF specific config so shouldn't be tied to VEGA GenFSK.
+config BT_CTLR_TIFS_HW
```
Probably @cvinayak can chip in here, but I'd guess it would make sense to add a 
depends on SOC_COMPATIBLE_NRF to the BT_CTLR_TIFS_HW option

alexandru-porosanu-nxp replied:
Well, since this is the Kconfig that's specific to the VEGABoard, I'd guess the depends for NRF SoCs is not needed, unless I'm missing something.

aescolar replied:
I meant it could be added in subsys/bluetooth/controller/Kconfig:560

alexandru-porosanu-nxp replied:
Oh, right; the reasoning behind doing it this way was I didn't want to pollute the controller Kconfig. Will change it accordingly.

cvinayak replied:
Yes (agree), the controller Kconfig be updated to hide options not applicable for other vendors

wopu-ot commented on subsys/bluetooth/controller/ll_sw/pdu.h:
```diff
@@ -527,9 +527,11 @@ struct pdu_data {
 #if !defined(CONFIG_BT_CTLR_DATA_LENGTH_CLEAR)
 	u8_t resv:8; /* TODO: remove nRF specific code */
 #endif /* !CONFIG_BT_CTLR_DATA_LENGTH_CLEAR */
+#endif /* !CONFIG_SOC_OPENISA_RV32M1_RISCV32 */
```
Maybe the TODO should be resolved instead of piling up #ifdefs?

alexandru-porosanu-nxp replied:
Agree with the removal. The reason it is still here is that we will probably use a similarly-named field when implementing LL encryption. But for now it makes no sense to keep it.

cvinayak replied:
This field is required by nRF platforms. Instead of the member, a macro can be placed that is replaced with this field by nordic's vendor headers. But this should be a separate PR task, and not to pollute this PR.

alexandru-porosanu-nxp replied:
@cvinayak , so I will leave these #defines as-is and send a new PR which implements the proposed solution for this field. Is this OK with you?

cvinayak replied:
@alexandru-porosanu-nxp yes, ok with me.

MaureenHelm commented on subsys/bluetooth/controller/ll_sw/openisa/README:
```diff
@@ -0,0 +1,23 @@
+BLE support on VEGAboard HW  Rev 1.0
+
+Currently the following sample applications are supported by the BLE SW LL
+and BLE host stack running on RI5CY core of RV32M1 SoC:
```
Move the README content to a new section in `boards/riscv/rv32m1_vega/doc/index.rst`

alexandru-porosanu-nxp replied:
Right, will do this, and also add the missing limitations

MaureenHelm reviewed (changes requested):
I'm pretty excited to see this! Thanks @alexandru-porosanu-nxp 

Need to modify west.yml to pick up the hal_openisa changes
```diff
diff --git a/west.yml b/west.yml
index 25d8890834..f8365bb79b 100644
--- a/west.yml
+++ b/west.yml
@@ -47,7 +47,7 @@ manifest:
       revision: 1b14177ff2176a1d17f3dd5e7e217f44337255db
       path: modules/hal/nordic
     - name: hal_openisa
-      revision: be5c01f86c96500def5079bcc58d2baefdffb6c8
+      revision: pull/1/head
       path: modules/hal/openisa
     - name: hal_microchip
       revision: 85302959c0c659311cf90ac51d133e5ce19c9288
```

Please add a test to `samples/bluetooth/peripheral_hr/sample.yaml` to ensure this builds in CI.

alexandru-porosanu-nxp committed:
ext: hal: openisa: sync with updates for BLE SW LL

There are a couple of commits that are needed for
the BLE SW LL to function properly on the VEGABoard:
- TRNG entropy driver
- GenFSK radio enablement + configuration

As such, make sure that the commit in the manifest
points to the appropriate SHA in the hal/openisa
repo.

Signed-off-by: Alex Porosanu <alexandru.porosanu@nxp.com>

alexandru-porosanu-nxp commented on subsys/bluetooth/controller/ll_sw/openisa/hal/RV32M1/radio/radio.c:
```diff
@@ -0,0 +1,1009 @@
+/*
```
You are correct, it's a similar radio IP, there are a couple of slight HW differences between the one on RV32M1 and the one on KW41Z. But for now, we'd like to focus on getting the SW LL running (and accepted) on this platform, and as a possible next step, enable other platforms f.i. KW41Z.

alexandru-porosanu-nxp committed:
bluetooth: kconfig: disable some options for openisa/RV32M1

Some Bluetooth options are currently not supported by the experimental
BLE SW LL implementation done on VEGABoard. As such, hide them from
the user altogether.

The full list of disabled config options is as follows:

- CONFIG_BT_PHY_UPDATE
- CONFIG_BT_DATA_LEN_UPDATE
- CONFIG_BT_HCI_VS
- CONFIG_BT_CTLR_LE_ENC
- CONFIG_BT_CTLR_CONN_PARAM_REQ
- CONFIG_BT_CTLR_EXT_REJ_IND
- CONFIG_BT_CTLR_SLAVE_FEAT_REQ
- CONFIG_BT_CTLR_LE_PING
- CONFIG_BT_CTLR_PRIVACY
- CONFIG_BT_CTLR_EXT_SCAN_FP
- CONFIG_BT_CTLR_CHAN_SEL_2
- CONFIG_BT_CTLR_ADV_EXT
- CONFIG_BT_CTLR_XTAL_ADVANCED
- CONFIG_BT_CTLR_SCHED_ADVANCED
- CONFIG_BT_CTLR_TIFS_HW

Signed-off-by: Alex Porosanu <alexandru.porosanu@nxp.com>

alexandru-porosanu-nxp committed:
samples: bluetooth: peripheral_hr: add VEGABoard sanity testing

Add a special target that's marked build_only so the BLE SW LL
implementation on VEGABoard is built daily. This helps ensuring
that it doesn't get inadvertedly broken by subsequent updates.
The name of the target is peripheral_hr_rv32m1_vega_ri5cy
and it can be run with the following command:

sanitycheck -v --all -p rv32m1_vega_ri5cy \
            -x=CMAKE_REQUIRED_FLAGS=-Wl,-dT=/dev/null \
            --test samples/bluetooth/peripheral_hr/\
              sample.bluetooth.peripheral_hr_rv32m1_vega_ri5cy

Note: The extra CMake flags are required if running with the official
      VEGABoard compiler, due to this linker issue:
      https://github.com/pulp-platform/pulpino/issues/240

Signed-off-by: Alex Porosanu <alexandru.porosanu@nxp.com>

dbkinder commented on boards/riscv/rv32m1_vega/doc/index.rst:
```diff
@@ -116,6 +116,44 @@ supports the following hardware features:
+   - **no power-save**
+   - no TX power adjustment
+
+.. note::
```
This note should just be part of the previous note, so delete this directive line.

dbkinder commented on boards/riscv/rv32m1_vega/doc/index.rst:
```diff
@@ -116,6 +116,44 @@ supports the following hardware features:
+- peripheral_ht (Health Thermometer)
+- scan_adv
+
+.. warning::
```
This feels more like a "note" than a "warning" (nothing of potential harm or damage is described here).

alexandru-porosanu-nxp replied:
Correct, the logic behind putting it as a 'warning' instead of 'note' was because some of the limitations are unexpected let's say (f.i. controller-based encryption). But you are right, nothing of potential harm is described, and as such I will make it a "note" and remove the previous "note" directive.

joerchan commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -239,14 +248,16 @@ if BT_CONN
 
 config BT_CTLR_LE_ENC
```
Maybe we should have an easier way for controllers to determine support for the different features.

How about
config BT_CTRL_LE_SUPPORT
  bool

config BT_CTLR_LE_ENC
  bool "LE Encryption
  depends on BT_CTRL_LE_ENC_SUPPORT
  default y

config BT_LLL_VENDOR_NORDIC
	bool "Use Nordic LLL"
        select BT_LL_ENC_SUPPORT

alexandru-porosanu-nxp replied:
Right, so all the various options would get set under the corresponding BT_LL_VENDOR_*, making things much more readable, instead of being scattered across all over the place (as I've done it). @cvinayak , what do you think? Should I go ahead and update the Kconfigs accordingly?

cvinayak replied:
Agree. Yes, please go ahead.

joerchan replied:
@alexandru-porosanu-nxp This PR is quite huge, if you prefer we can take the suggested change as separate PR afterwards.

alexandru-porosanu-nxp replied:
Actually, I went ahead and done it anyway. It should be there in the latest version of the PR. So @joerchan , if you could have a look and let me know if it looks OK, that would be great.

joerchan replied:
Thanks. This looks much more managable. The last 3 features I think are very NRF specific. And I think maybe should be moved under LLL_VENDOR_NORDIC as they are closely related to NRF hardware features.

cvinayak reviewed (approved):
Controller related changes LGTM. Any non-critical change requests can be addressed post merge of this PR.

dbkinder commented on boards/riscv/rv32m1_vega/doc/index.rst:
```diff
@@ -116,6 +116,42 @@ supports the following hardware features:
+==================================================
+This is an experimental feature supported on the Zephyr's RI5CY
+configuration, ``rv32m1_vega_ri5cy``. It  uses the Software Link Layer
+framework by Nordic Semi to enable the usage of the on-SoC radio and
```
better would be:
```
framework by Nordic Semi to enable the on-SoC radio and
transceiver for implementing a software-defined BLE controller.
```

alexandru-porosanu-nxp replied:
OK, will change tand re-push.

dbkinder commented on boards/riscv/rv32m1_vega/doc/index.rst:
```diff
@@ -116,6 +116,42 @@ supports the following hardware features:
+   - no TX power adjustment
+
+   Limitations marked **bold** above are planned to be addressed on this
+   platform
```
add a period at the end of the sentence.

alexandru-porosanu-nxp replied:
Missed this one, thanks!

dbkinder reviewed (changes requested):
a couple of tweaks to the docs.

joerchan commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -7,6 +7,56 @@
+config BT_CTLR_ADV_EXT_SUPPORT
+	bool
+
+config BT_CTLR_XTAL_ADVANCED_SUPPORT
```
@cvinayak These last 3 seems very nordic specific, should these perhaps be moved under LLL_VENDOR_NORDIC?

cvinayak replied:
Agree.

alexandru-porosanu-nxp replied:
I can address them as part of a PR if @cvinayak agrees.

joerchan commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -7,6 +7,56 @@
 comment "BLE Controller support"
 
+# The following symbols are enabled depending if the controller actually
+# supports the respective features. In case of Nordic SW LL, all of these are
```
Your second sentence here is bound to be outdated. I would remove it. The information is maintained by the list of selects that the different LLL controller has.

alexandru-porosanu-nxp replied:
OK, will remove

joerchan commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -239,14 +312,16 @@ if BT_CONN
 config BT_CTLR_LE_ENC
 	bool "LE Encryption"
-	depends on !BT_CTLR_DATA_LENGTH_CLEAR && !BT_CTLR_PHY_2M_NRF
+	depends on !BT_CTLR_DATA_LENGTH_CLEAR && !BT_CTLR_PHY_2M_NRF && \
```
Is it possible to move the PHY_2M_NRF to where LLL_VENDOR_NORDIC selects ENC_SUPPORT?

`select BT_CTRL_LE_ENC_SUPPORT if BT_CTLR_PHY_2M_NRF`?

alexandru-porosanu-nxp replied:
Yeah, I think that's doable. I will include this in the next push

alexandru-porosanu-nxp replied:
@joerchan , there's something that seems amiss here:  subsys/bluetooth/controller/Kconfig on master contains the following:

`config BT_CTLR_LE_ENC`
`	depends on !BT_CTLR_DATA_LENGTH_CLEAR && !BT_CTLR_PHY_2M_NRF`

The proposed change in this PR is:

`config BT_CTLR_LE_ENC `
`	depends on !BT_CTLR_DATA_LENGTH_CLEAR && BT_CTLR_LE_ENC_SUPPORT`

So I guess that your suggestion above 

`select BT_CTRL_LE_ENC_SUPPORT if BT_CTLR_PHY_2M_NRF` 

should actually be 

`select BT_CTRL_LE_ENC_SUPPORT if !BT_CTLR_PHY_2M_NRF`

Or alternatively, rewrite BT_CTRL_LE_ENC to 

`config BT_CTLR_LE_ENC`
`	depends on !BT_CTLR_DATA_LENGTH_CLEAR && !BT_CTLR_LE_ENC_SUPPORT`

The reason that I've circled back to this is that the BabbleSim tests seem broken and by bisecting the commits, I've reached the one that modifies the Kconfig for controller. Comparing the generated .config on master with the one from this PR singled out the BT_CTRL_LE_ENC config options.

Please let me know if this makes sense.

rugeGerritsen commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -7,6 +7,56 @@
+# supports the respective features. In case of Nordic SW LL, all of these are
+# supported; for OpenISA SW LL, they're not
+
+config BT_LL_SW_LEGACY_SUPPORT
```
```BT_LL_SW_LEGACY_SUPPORT``` is only used in one occurence below. It would remove it and change 
```depends on BT_LL_SW_LEGACY_SUPPORT -> depends on SOC_COMPATIBLE_NRF```

alexandru-porosanu-nxp replied:
Right, since it's the only controller that supports the legacy mode, and this will be deprecated. So there's a really low probability another SW LL implementation will show up, supporting the legacy mode. As such, I will do the change.

rugeGerritsen commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -7,6 +7,56 @@
+config BT_LL_SW_LEGACY_SUPPORT
+	bool
+	depends on SOC_COMPATIBLE_NRF
+
```
Can we reorder the features to match the ordering in Table 4.5 in Core v5.1, Vol 6, Part B, Section 4.6?

alexandru-porosanu-nxp replied:
Sure. I've reordered them, and I will re-push. Do let me know if it looks OK.

dbkinder reviewed (approved):
+1 for doc changes, thanks!

rugeGerritsen commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -7,6 +7,48 @@
 
+# The following symbols are enabled depending if the controller actually
+# supports the respective features.
+
```
Maybe we should add `BT_CTRL_SLAVE_SUPPORT` and `BT_CTRL_MASTER_SUPPORT` as well?This will allow us to add ifdefs for master and slave specific code

alexandru-porosanu-nxp replied:
Yes, I think it makes sense. But I'd rather do this as a separate PR, after the current one gets merged. Is this acceptable?

wopu-ot replied:
What's wrong with using the already existing BT_CENTRAL/BT_PERIPHERAL?

rugeGerritsen replied:
I see your point. This CENTRAL/PERIPHERAL may be sufficient. CENTRAL/PERIPHERAL is host terminology, while MASTER/SLAVE is controller terminology

joerchan replied:
CENTRAL/PERIPHERAL are part of the common Kconfig variables for host and controller.
The question is whether we should add _SUPPORT kconfig options for them. And for those it would make sense to use controller MASTER/SLAVE terminology. So that it is possible to add controller with only SLAVE support for example.

wopu-ot replied:
I do not see the use case for having a separate MASTER/SLAVE configuration flag for the controller. If the controller supports only SLAVE, then compiling the host with CENTRAL support would not make a lot of sense.

George-Stefan replied:
Is this discussion relevant for Vega support?

joerchan replied:
It's relevant to the _SUPPORT change that was done as part of this PR.
But alexandru suggested to do this change separately and we agreed.. So we can end this discussion here.

dleach02 commented:
@aescolar , @wopu-ot , @joerchan  , @rugeGerritsen, 

Can we get some approvals (green check marks) for this PR based on the feedback to your comments?

George-Stefan commented on boards/riscv/rv32m1_vega/doc/index.rst:
```diff
@@ -116,6 +116,42 @@ supports the following hardware features:
+   - no TX power adjustment
+
+   Limitations marked **bold** above are planned to be addressed on this
+   platform
```
https://github.com/zephyrproject-rtos/zephyr/issues/19685

I can reproduce this bug on Vega all the time

George-Stefan commented on subsys/bluetooth/controller/ll_sw/openisa/lll/lll_adv.c:
```diff
@@ -0,0 +1,871 @@
+	    IS_ENABLED(CONFIG_BT_CTLR_GPIO_LNA_PIN)) {
+		radio_gpio_pa_lna_disable();
+	}
+	/* TODO: MOVE ^^ */
```
Also in Nordic lll_adv.c

aescolar commented:
@alexandru-porosanu-nxp Note that a PR cannot be merged until all automatically reported issues are fixed and all automated checks pass.
Reviewers tend to have an expectation that all those issues will be fixed promptly, and will tend to understand that if a PR has several failures in that regard it is not yet ready for review.

alexandru-porosanu-nxp commented:
> @alexandru-porosanu-nxp Note that a PR cannot be merged until all automatically reported issues are fixed and all automated checks pass.
> Reviewers tend to have an expectation that all those issues will be fixed promptly, and will tend to understand that if a PR has several failures in that regard it is not yet ready for review.

OK, understood. I just wanted to make sure that the PR contents make sense from the BLE LL SW owners' point of view and it doesn't break something. Once the owners agree with the form of the PR, it's quite easy to fix the automated checks (chckpatch, license, codeowners)

I've done some fixes and I have a couple of questions:

1. Regarding the codeowner checks, I see there is no codeowner added for /drivers/entropy. As such, I've taken the 3 most active commiters and added them as owners. Does this make sense?

2. Regarding the checkpatch warning, that the openisa string vendor  appears undocumented, I see there are plenty of other nodes that use the same vendor, but it is not added to the vendor-prefixes.txt file. I've tentatively added a separate patch that adds an entry to vendor-prefixes, linking openisa vendor to the OpenISA org. Not sure if this right.

It seems that the Kconfig update broke something on the Nordic side. It seems to be related to the LE encryption support. @joerchan suggested that BT_CTLR_LE_ENC_SUPPORT should get selected if BT_CTLR_PHY_2M_NRF is selected...but this seems to break all (Nordic) controllers not having BT_CTLR_PHY_2M_NRF set. I'm probably missing something here so I'll continue to have a look.

aescolar commented:
>I just wanted to make sure..

That is fair. My response was triggered due to David's request for approvals. ( Anyhow, approvals on PRs that are still changing considerably are kind of a funny thing, as all further changes would not be necessarily reviewed )

> Regarding the codeowner checks, I see there is no codeowner added for /drivers/entropy. As such, I've taken the 3 most active commiters and added them as owners. Does this make sense?

Not so fast :) , those people would need to add themselves or at least consent to it. Given that this PR is only having an issue with 2 new files you add, you may want to just take responsibility of these
drivers/entropy/Kconfig.rv32m1
drivers/entropy/entropy_rv32m1_trng.c

so add a line like:
drivers/entropy/*rv32m1*   @mhelm 
or whoever from NXP agrees to support it

alexandru-porosanu-nxp commented:
> > I just wanted to make sure..
> 
> That is fair. My response was triggered due to David's request for approvals. ( Anyhow, approvals on PRs that are still changing considerably are kind of a funny thing, as all further changes would not be necessarily reviewed )
> 
> > Regarding the codeowner checks, I see there is no codeowner added for /drivers/entropy. As such, I've taken the 3 most active commiters and added them as owners. Does this make sense?
> 
> Not so fast :) , those people would need to add themselves or at least consent to it. Given that this PR is only having an issue with 2 new files you add, you may want to just take responsibility of these
> drivers/entropy/Kconfig.rv32m1
> drivers/entropy/entropy_rv32m1_trng.c
> 
> so add a line like:
> `drivers/entropy/*rv32m1* @mhelm `
> or whoever from NXP agrees to support it

OK, got it. Will change accordingly. I will re-push once Joakim agrees with the proposed change (or suggests an alternative)

joerchan commented:
> It seems that the Kconfig update broke something on the Nordic side. It seems to be related to the LE encryption support. @joerchan suggested that BT_CTLR_LE_ENC_SUPPORT should get selected if BT_CTLR_PHY_2M_NRF is selected...but this seems to break all (Nordic) controllers not having BT_CTLR_PHY_2M_NRF set. I'm probably missing something here so I'll continue to have a look.

Did you fix this? It should be !BT_CTLR_PHY_2M_NRF, which seems to be the case now.
I couldn't see anything in shippable that pointed to a problem here, if there is could you point me to it?

The overlay files in all the samples looks like a maintenance problem. But I can understand if you prefer to handle this separately. I don't know anything about DTS so I cannot really help you. I guess @galak would be the one to assist.

There seems to be a lot of DTS overlay failures in Shippable, please have a look.

alexandru-porosanu-nxp commented:
> > It seems that the Kconfig update broke something on the Nordic side. It seems to be related to the LE encryption support. @joerchan suggested that BT_CTLR_LE_ENC_SUPPORT should get selected if BT_CTLR_PHY_2M_NRF is selected...but this seems to break all (Nordic) controllers not having BT_CTLR_PHY_2M_NRF set. I'm probably missing something here so I'll continue to have a look.
> 
> @alexandru-porosanu-nxp
> Did you fix this? It should be !BT_CTLR_PHY_2M_NRF, which seems to be the case now.
> I couldn't see anything in shippable that pointed to a problem here, if there is could you point me to it?

Yes, so I've pushed the fix for the 2M PHY selection but I broke the Nordic legacy SW LL. This is broken since the first push, since no *_SUPPORT values get defined. Currently I have a fix for this and I will re-push, so I'd kindly ask you to have a look again.

> 
> The overlay files in all the samples looks like a maintenance problem. But I can understand if you prefer to handle this separately. I don't know anything about DTS so I cannot really help you. I guess @galak would be the one to assist.

I agree it's messy. So once this PR is merged, I will send a proposal to address the scattering of per-sample overlays. 

> 
> There seems to be a lot of DTS overlay failures in Shippable, please have a look.

Yes, I broke the zero RISCY core builds with the Generic FSK addition. It is an easy fix, which I've done and I'll re-push soon.

joerchan commented:
@alexandru-porosanu-nxp Ok. Then I understand, looks like you have fixed it. 👍

joerchan reviewed (approved):
This looks good to me, with the assumption that the DTS overlay files will be cleaned up later.

alexandru-porosanu-nxp commented:
> @alexandru-porosanu-nxp Ok. Then I understand, looks like you have fixed it. 👍

Yep, and now all checks are passing. Thanks for looking into this

alexandru-porosanu-nxp commented:
Rebase + fix after ticker changes in #20254

alexandru-porosanu-nxp commented:
Adjusted EVENT_TICKER_RES_MARGIN_US for OpenISA SW LL (should be 64<<1)

alexandru-porosanu-nxp commented:
Fixed conflicts after the merge of #18557 
Fixed compilation failures in OpenISA SW LL after the merge of #20233 
Needs #20307 to be merged, central_hr doesn't compile now (BabbleSim tests on the other hand should pass)

alexandru-porosanu-nxp commented:
Rebase after #20307 merged into master.

alexandru-porosanu-nxp commented:
Rebased on latest master, fixed conflicts after #17736

alexandru-porosanu-nxp commented:
Removed some leftovers from the legacy LL

alexandru-porosanu-nxp commented:
Rebased on latest master, removed some more leftovers from the legacy LL

alexandru-porosanu-nxp commented:
Rebased on latest master and resolved conflict after #19712 merge

alexandru-porosanu-nxp commented:
Rebased on latest master and updated west.yml after the merge of https://github.com/zephyrproject-rtos/hal_openisa/pull/1

joerchan commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -237,14 +320,15 @@ if BT_CONN
 config BT_CTLR_LE_ENC
 	bool "LE Encryption"
-	depends on !BT_CTLR_DATA_LENGTH_CLEAR && !BT_CTLR_PHY_2M_NRF
+	depends on !BT_CTLR_DATA_LENGTH_CLEAR && BT_CTLR_LE_ENC_SUPPORT
```
`depends on BT_CTLR_LE_ENC_SUPPORT`

alexandru-porosanu-nxp replied:
Done

joerchan commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -274,7 +360,7 @@ config BT_CTLR_LE_PING
 config BT_CTLR_PRIVACY
 	bool "LE Controller-based Privacy"
-	depends on !SOC_SERIES_NRF51X
+	depends on !SOC_SERIES_NRF51X && BT_CTLR_PRIVACY_SUPPORT
```
`depends on BT_CTLR_PRIVACY_SUPPORT`

alexandru-porosanu-nxp replied:
Done

joerchan commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -44,10 +103,34 @@ config BT_LLL_VENDOR_NORDIC
+	select BT_CTLR_CONN_PARAM_REQ_SUPPORT
+	select BT_CTLR_EXT_REJ_IND_SUPPORT
+	select BT_CTLR_SLAVE_FEAT_REQ_SUPPORT
+	select BT_DATA_LEN_UPDATE_SUPPORT
```
```
select BT_CTRL_DATA_LEN_UPDATE_SUPPORT if !SOC_SERIES_NRF51X || BT_CTLR_DATA_LENGTH_CLEAR
select BT_CTLR_PRIVACY_SUPPORT if !SOC_SERIES_NRF51X
select BT_CTLR_EXT_SCAN_FP_SUPPORT
select BT_CTLR_PHY_UPDATE_SUPPORT if !SOC_SERIES_NRF51X || BT_CTLR_PHY_2M_NRF
```

alexandru-porosanu-nxp replied:
Done

alexandru-porosanu-nxp replied:
> ```
> select BT_CTRL_DATA_LEN_UPDATE_SUPPORT if !SOC_SERIES_NRF51X || BT_CTLR_DATA_LENGTH_CLEAR
> select BT_CTLR_PRIVACY_SUPPORT if !SOC_SERIES_NRF51X
> select BT_CTLR_EXT_SCAN_FP_SUPPORT
> select BT_CTLR_PHY_UPDATE_SUPPORT if !SOC_SERIES_NRF51X || BT_CTLR_PHY_2M_NRF
> ```

I think that the BT_CTLR_PHY_UPDATE_SUPPORT is always selected, since if SOC_SERIES_NRF51X  = 0, since BT_CTLR_PHY_2M_NRF depends on SOC_SERIES_NRF51X, the logical value of the if is 1 (!0 || 0 == 1); the same goes for the opposite (if SOC_SERIES_NRF51X  == 1, then !1 || 1 == 1).

So, @joerchan, I think this should remain as is. What do you think?

joerchan commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -44,10 +103,34 @@ config BT_LLL_VENDOR_NORDIC
 	select ENTROPY_NRF5_RNG
 	select ENTROPY_NRF5_BIAS_CORRECTION
+
+	select BT_CTLR_LE_ENC_SUPPORT if !BT_CTLR_PHY_2M_NRF
```
`select BT_CTLR_LE_ENC_SUPPORT if !BT_CTLR_DATA_LENGTH_CLEAR && !BT_CTLR_PHY_2M_NRF`

joerchan commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -5,6 +5,48 @@
+config BT_CTLR_SLAVE_FEAT_REQ_SUPPORT
+	bool
+
+config BT_DATA_LEN_UPDATE_SUPPORT
```
`BT_CTLR_DATA_LEN_UPDATE_SUPPORT`

alexandru-porosanu-nxp replied:
Done

joerchan commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -5,6 +5,48 @@
+config BT_CTLR_EXT_SCAN_FP_SUPPORT
+	bool
+
+config BT_PHY_UPDATE_SUPPORT
```
`BT_CTLR_PHY_UPDATE_SUPPORT`

alexandru-porosanu-nxp replied:
Done

joerchan commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -293,7 +379,7 @@ config BT_CTLR_RL_SIZE
+	depends on BT_OBSERVER && BT_CTLR_EXT_SCAN_FP_SUPPORT
 	default y
 	help
 	  Enable support for Bluetooth v4.2 LE Extended Scanner Filter Policies
```
I can't comment on lines not part of the diff ...

```
config BT_CTLR_DATA_LENGTH
	# Hidden option to enable support for Bluetooth v4.2 LE Data Length
	# Update procedure in the Controller.
	bool
	depends on BT_DATA_LEN_UPDATE && BT_CTLR_DATA_LENGTH_UPDATE_SUPPORT
	default y
```

and

```
config BT_CTLR_PHY
	# Hidden option to enable support for Bluetooth 5.0 PHY Update
	# Procedure in the Controller.
	bool
	depends on BT_PHY_UPDATE && BT_CTLR_PHY_UPDATE_SUPPORT
	select BT_CTLR_EXT_REJ_IND
	default y
```

We might also need `_PHY_2M_SUPPORTED` and `_PHY_CODED_SUPPORTED`

alexandru-porosanu-nxp replied:
> Sorry to add these comments so late. I missed it earlier.

That's OK, they're not really hard to address. As such, I'll run our internal regression suite and then re-push to ensure that I haven't broken something else.

Anyhow, for the 2M_SUPPORTED and CODED_SUPPORTED, I think that should be addressed in a separate PR, does this make sense to you?

joerchan reviewed (changes requested):
Sorry to add these comments so late. I missed it earlier.

joerchan reviewed (dismissed):
> Anyhow, for the 2M_SUPPORTED and CODED_SUPPORTED, I think that should be addressed in a separate PR, does this make sense to you?

That makes sense to me. Thank you for addressing the issues.
There are a few subtle `CTLR` `CTRL` issues. Done that one many times myself :)

alexandru-porosanu-nxp commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -33,6 +75,26 @@ config BT_LL_SW_LEGACY
+	select BT_CTLR_SLAVE_FEAT_REQ_SUPPORT
+	select BT_CTRL_DATA_LEN_UPDATE_SUPPORT if !SOC_SERIES_NRF51X || \
+						  BT_CTLR_DATA_LENGTH_CLEAR
+	select BT_CTLR_PRIVACY_SUPPORT
```
Right, missed that one, thanks!

alexandru-porosanu-nxp commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -328,7 +420,7 @@ config BT_CTLR_PHY
-	depends on BT_PHY_UPDATE
+	depends on BT_PHY_UPDATE && BT_CTLR_PHY_UPDATE_SUPPORT
 	select BT_CTLR_EXT_REJ_IND
 	default y if SOC_COMPATIBLE_NRF52X || BT_CTLR_PHY_2M_NRF
```
Done

alexandru-porosanu-nxp commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -313,7 +405,7 @@ config BT_CTLR_DATA_LENGTH
 	bool
-	depends on BT_DATA_LEN_UPDATE
+	depends on BT_DATA_LEN_UPDATE && BT_CTLR_DATA_LENGTH_UPDATE_SUPPORT
 	default y if SOC_COMPATIBLE_NRF52X || BT_CTLR_DATA_LENGTH_CLEAR
```
Done

alexandru-porosanu-nxp commented:
> > Anyhow, for the 2M_SUPPORTED and CODED_SUPPORTED, I think that should be addressed in a separate PR, does this make sense to you?
> 
> That makes sense to me. Thank you for addressing the issues.
> There are a few subtle `CTLR` `CTRL` issues. Done that one many times myself :)

Yes, it seems I swapped them a bit. I'm getting a dependency loop now for BT_PHY_UPDATE, so it'll be short while till I push again.

alexandru-porosanu-nxp commented:
Rebased on latest master, resolved conflicts

joerchan commented:
> Yes, it seems I swapped them a bit. I'm getting a dependency loop now for BT_PHY_UPDATE, so it'll be short while till I push again.

I think a better solution to the BT_PHY_UPDATE loop is this:
https://github.com/joerchan/zephyr/commit/e9b04bf08cb9d15900df4f2fe464d92ac2299773

This is similar to have the Cleartext Data Length exception works.

alexandru-porosanu-nxp commented:
> > Yes, it seems I swapped them a bit. I'm getting a dependency loop now for BT_PHY_UPDATE, so it'll be short while till I push again.
> 
> I think a better solution to the BT_PHY_UPDATE loop is this:
> [joerchan@e9b04bf](https://github.com/joerchan/zephyr/commit/e9b04bf08cb9d15900df4f2fe464d92ac2299773)
> 
> This is similar to have the Cleartext Data Length exception works.

OK, will integrate

joerchan reviewed (approved):
Great job. Thank you.

carlescufi commented:
Merging since CI failure was a one-off and we know this passes
